base/math.jl

```diff
@@ -349,8 +349,45 @@ julia> ldexp(5., 2)
 20.0
 ```
 """
-ldexp(x::Float64,e::Integer) = ccall((:scalbn,libm),  Float64, (Float64,Int32), x, Int32(e))
-ldexp(x::Float32,e::Integer) = ccall((:scalbnf,libm), Float32, (Float32,Int32), x, Int32(e))
+function ldexp{T<:AbstractFloat}(x::T, q::Integer)
+    n = Int(q)
+    xu = reinterpret(Unsigned, x)
+    xs = xu & ~sign_mask(T)
+    xs >= exponent_mask(T) && return x # NaN or Inf
+    k = Int(xs >> significand_bits(T))
+    if k == 0 # x is subnormal
+        xs == 0 && return x # +-0
+        m = unsigned(leading_zeros(xs) - exponent_bits(T))
+        xs <<= m
+        xu = xs | (xu & sign_mask(T))
+        k = 1 - signed(m)
+        if n < -50000 # underflow
+            # otherwise may have have an integer underflow in the following k += n computation
+            return flipsign(T(0.0), x)
+        end
+    end
+    k += n
```

Comment:
In that case, another option would be
```julia
if q > typemax(Int)
    return flipsign(T(Inf), x)
elseif q < typemin(Int)
    return flipsign(T(0), x)
end
n = q % Int
```
which should be able to optimise away the check if `q::Int`.